research/a3c_blogpost/a3c_cartpole.py

```diff
@@ -0,0 +1,390 @@
+import os
+os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"   # see issue #152
+os.environ["CUDA_VISIBLE_DEVICES"] = ""
+
+
+import threading
+import gym
+import multiprocessing
+import numpy as np
+from queue import Queue
+import argparse
+import matplotlib.pyplot as plt
+
+
+import tensorflow as tf
+from tensorflow.python import keras
+from tensorflow.python.keras import layers
+
+tf.enable_eager_execution()
+print("Eager execution: {}".format(tf.executing_eagerly()))
+
+def str2bool(v):
+  if v.lower() in ('yes', 'true', 't', 'y', '1'):
+    return True
+  elif v.lower() in ('no', 'false', 'f', 'n', '0'):
+    return False
+  else:
+    raise argparse.ArgumentTypeError('Boolean value expected.')
+
+
+parser = argparse.ArgumentParser(description='Run A3C algorithm on the game '
+                                             'Cartpole.')
+parser.add_argument('--algorithm', default='a3c', type=str,
+                    help='Choose between \'a3c\' and \'random\'.')
+parser.add_argument("--train", type=str2bool, default=True,
+                    help='Train our model or to run an existing model and '
+                         'watch it play.')
+parser.add_argument('--lr', default=0.0005,
+                    help='Learning rate for the shared optimizer.')
+parser.add_argument('--update-freq', default=20, type=int,
+                    help='How often to update the global model.')
+parser.add_argument('--max-eps', default=2000, type=int,
+                    help='Global maximum number of episodes to run.')
+parser.add_argument('--gamma', default=0.99,
+                    help='Discount factor of rewards.')
+parser.add_argument('--save-dir', default='/tmp/', type=str,
+                    help='Directory in which you desire to save the model.')
+
+args = parser.parse_args()
+
+
+class ActorCriticModel(keras.Model):
+  def __init__(self, state_size, action_size):
+    super(ActorCriticModel, self).__init__()
+    self.state_size = state_size
+    self.action_size = action_size
+    self.dense1 = layers.Dense(100, activation='relu')
+    self.policy_logits = layers.Dense(action_size)
+    self.dense2 = layers.Dense(100, activation='relu')
+    self.values = layers.Dense(1)
+
+  def call(self, inputs):
+    # Forward pass
+    x = self.dense1(inputs)
+    logits = self.policy_logits(x)
+    v1 = self.dense2(inputs)
+    values = self.values(v1)
+    return logits, values
+
+
+def tf_wrap(np_array, dtype=np.float32):
```

Comment:
Do you need this? I think a tf.cast(np_array, tf.dtype) should do the trick if a tf.convert_to_tensor(np_array, tf_dtype) doesn't work